File: python/src/falcon_core/physics/config/core/group.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.group import Group as _CGroup
from falcon_core.autotuner_interfaces.names.channel import Channel
from falcon_core.physics.device_structures.connection import Connection
from falcon_core.physics.device_structures.connections import Connections
from falcon_core.physics.config.geometries.gate_geometry_array1_d import GateGeometryArray1D

class Group:
    """Python wrapper for Group."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def from_json(cls, json: str) -> Group:
        return cls(_CGroup.from_json(json))

    @classmethod
    def new(cls, name: Channel, num_dots: Any, screening_gates: Connections, reservoir_gates: Connections, plunger_gates: Connections, barrier_gates: Connections, order: Connections) -> Group:
        obj = cls(_CGroup.new(name._c if name is not None else None, num_dots, screening_gates._c if screening_gates is not None else None, reservoir_gates._c if reservoir_gates is not None else None, plunger_gates._c if plunger_gates is not None else None, barrier_gates._c if barrier_gates is not None else None, order._c if order is not None else None))
        obj._ref_name = name  # Keep reference alive
        obj._ref_screening_gates = screening_gates  # Keep reference alive
        obj._ref_reservoir_gates = reservoir_gates  # Keep reference alive
        obj._ref_plunger_gates = plunger_gates  # Keep reference alive
        obj._ref_barrier_gates = barrier_gates  # Keep reference alive
        obj._ref_order = order  # Keep reference alive
        return obj

    def copy(self, ) -> Group:
        ret = self._c.copy()
        return Group._from_capi(ret)

    def equal(self, other: Group) -> bool:
        ret = self._c.equal(other._c if other is not None else None)
        return ret

    def not_equal(self, other: Group) -> bool:
        ret = self._c.not_equal(other._c if other is not None else None)
        return ret

    def to_json(self, ) -> str:
        ret = self._c.to_json()
        return ret

    def num_dots(self, ) -> int:
        ret = self._c.num_dots()
        return ret

    def order(self, ) -> GateGeometryArray1D:
        ret = self._c.order()
        if ret is None: return None
        return GateGeometryArray1D._from_capi(ret)

    def has_channel(self, channel: Channel) -> bool:
        ret = self._c.has_channel(channel._c if channel is not None else None)
        return ret

    def is_charge_sensor(self, ) -> bool:
        ret = self._c.is_charge_sensor()
        return ret

    def has_ohmic(self, ohmic: Connection) -> bool:
        ret = self._c.has_ohmic(ohmic._c if ohmic is not None else None)
        return ret

    def has_gate(self, gate: Connection) -> bool:
        ret = self._c.has_gate(gate._c if gate is not None else None)
        return ret

    def has_barrier_gate(self, barrier_gate: Connection) -> bool:
        ret = self._c.has_barrier_gate(barrier_gate._c if barrier_gate is not None else None)
        return ret

    def has_plunger_gate(self, plunger_gate: Connection) -> bool:
        ret = self._c.has_plunger_gate(plunger_gate._c if plunger_gate is not None else None)
        return ret

    def has_reservoir_gate(self, reservoir_gate: Connection) -> bool:
        ret = self._c.has_reservoir_gate(reservoir_gate._c if reservoir_gate is not None else None)
        return ret

    def has_screening_gate(self, screening_gate: Connection) -> bool:
        ret = self._c.has_screening_gate(screening_gate._c if screening_gate is not None else None)
        return ret

    @property
    def all_channel_gates(self) -> Connections:
        ret = self._c.get_all_channel_gates()
        if ret is None: return None
        return Connections._from_capi(ret)

    @property
    def all_connections(self) -> Connections:
        ret = self._c.get_all_connections()
        if ret is None: return None
        return Connections._from_capi(ret)

    @property
    def all_gates(self) -> Connections:
        ret = self._c.get_all_gates()
        if ret is None: return None
        return Connections._from_capi(ret)

    @property
    def barrier_gate(self) -> Connection:
        ret = self._c.get_barrier_gate()
        if ret is None: return None
        return Connection._from_capi(ret)

    @property
    def barrier_gates(self) -> Connections:
        ret = self._c.barrier_gates()
        if ret is None: return None
        return Connections._from_capi(ret)

    @property
    def dot_gate(self) -> Connection:
        ret = self._c.get_dot_gate()
        if ret is None: return None
        return Connection._from_capi(ret)

    @property
    def dot_gates(self) -> Connections:
        ret = self._c.dot_gates()
        if ret is None: return None
        return Connections._from_capi(ret)

    @property
    def gate(self) -> Connection:
        ret = self._c.get_gate()
        if ret is None: return None
        return Connection._from_capi(ret)

    @property
    def name(self) -> Channel:
        ret = self._c.name()
        if ret is None: return None
        return Channel._from_capi(ret)

    @property
    def ohmic(self) -> Connection:
        ret = self._c.get_ohmic()
        if ret is None: return None
        return Connection._from_capi(ret)

    @property
    def ohmics(self) -> Connections:
        ret = self._c.ohmics()
        if ret is None: return None
        return Connections._from_capi(ret)

    @property
    def plunger_gate(self) -> Connection:
        ret = self._c.get_plunger_gate()
        if ret is None: return None
        return Connection._from_capi(ret)

    @property
    def plunger_gates(self) -> Connections:
        ret = self._c.plunger_gates()
        if ret is None: return None
        return Connections._from_capi(ret)

    @property
    def reservoir_gate(self) -> Connection:
        ret = self._c.get_reservoir_gate()
        if ret is None: return None
        return Connection._from_capi(ret)

    @property
    def reservoir_gates(self) -> Connections:
        ret = self._c.reservoir_gates()
        if ret is None: return None
        return Connections._from_capi(ret)

    @property
    def screening_gate(self) -> Connection:
        ret = self._c.get_screening_gate()
        if ret is None: return None
        return Connection._from_capi(ret)

    @property
    def screening_gates(self) -> Connections:
        ret = self._c.screening_gates()
        if ret is None: return None
        return Connections._from_capi(ret)

    def __repr__(self):
        return f"Group({self.to_json()})"

    def __str__(self):
        return self.to_json()

    def __hash__(self):
        """Hash based on JSON representation"""
        return hash(self.to_json())

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, Group):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, Group):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/physics/config/core/group.ml
open Ctypes
open Capi_bindings
open ErrorHandling

open Falcon_core.Autotuner_interfaces.Names
open Falcon_core.Physics.Config.Geometries
open Falcon_core.Physics.Device_structures

class c_group (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.group_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module Group = struct
  type t = c_group

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.group_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_group ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.group_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_group ptr

  let make (name : Channel.t) (num_dots : int) (screening_gates : Connections.t) (reservoir_gates : Connections.t) (plunger_gates : Connections.t) (barrier_gates : Connections.t) (order : Connections.t) : t =
    ErrorHandling.multi_read [name; screening_gates; reservoir_gates; plunger_gates; barrier_gates; order] (fun () ->
      let ptr = Capi_bindings.group_create name#raw num_dots screening_gates#raw reservoir_gates#raw plunger_gates#raw barrier_gates#raw order#raw in
      ErrorHandling.raise_if_error ();
      new c_group ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.group_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.group_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let name (handle : t) : Channel.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_name handle#raw in
      ErrorHandling.raise_if_error ();
      new c_channel result
    )

  let numDots (handle : t) : int =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_num_dots handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let order (handle : t) : GateGeometryArray1D.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_order handle#raw in
      ErrorHandling.raise_if_error ();
      new c_gategeometryarray1d result
    )

  let hasChannel (handle : t) (channel : Channel.t) : bool =
    ErrorHandling.multi_read [handle; channel] (fun () ->
      let result = Capi_bindings.group_has_channel handle#raw channel#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let isChargeSensor (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_is_charge_sensor handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let getAllChannelGates (handle : t) : Connections.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_get_all_channel_gates handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connections result
    )

  let screeningGates (handle : t) : Connections.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_screening_gates handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connections result
    )

  let reservoirGates (handle : t) : Connections.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_reservoir_gates handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connections result
    )

  let plungerGates (handle : t) : Connections.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_plunger_gates handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connections result
    )

  let barrierGates (handle : t) : Connections.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_barrier_gates handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connections result
    )

  let ohmics (handle : t) : Connections.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_ohmics handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connections result
    )

  let dotGates (handle : t) : Connections.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_dot_gates handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connections result
    )

  let getOhmic (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_get_ohmic handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let getBarrierGate (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_get_barrier_gate handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let getPlungerGate (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_get_plunger_gate handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let getReservoirGate (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_get_reservoir_gate handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let getScreeningGate (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_get_screening_gate handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let getDotGate (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_get_dot_gate handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let getGate (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_get_gate handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let getAllGates (handle : t) : Connections.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_get_all_gates handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connections result
    )

  let getAllConnections (handle : t) : Connections.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.group_get_all_connections handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connections result
    )

  let hasOhmic (handle : t) (ohmic : Connection.t) : bool =
    ErrorHandling.multi_read [handle; ohmic] (fun () ->
      let result = Capi_bindings.group_has_ohmic handle#raw ohmic#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let hasGate (handle : t) (gate : Connection.t) : bool =
    ErrorHandling.multi_read [handle; gate] (fun () ->
      let result = Capi_bindings.group_has_gate handle#raw gate#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let hasBarrierGate (handle : t) (barrier_gate : Connection.t) : bool =
    ErrorHandling.multi_read [handle; barrier_gate] (fun () ->
      let result = Capi_bindings.group_has_barrier_gate handle#raw barrier_gate#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let hasPlungerGate (handle : t) (plunger_gate : Connection.t) : bool =
    ErrorHandling.multi_read [handle; plunger_gate] (fun () ->
      let result = Capi_bindings.group_has_plunger_gate handle#raw plunger_gate#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let hasReservoirGate (handle : t) (reservoir_gate : Connection.t) : bool =
    ErrorHandling.multi_read [handle; reservoir_gate] (fun () ->
      let result = Capi_bindings.group_has_reservoir_gate handle#raw reservoir_gate#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let hasScreeningGate (handle : t) (screening_gate : Connection.t) : bool =
    ErrorHandling.multi_read [handle; screening_gate] (fun () ->
      let result = Capi_bindings.group_has_screening_gate handle#raw screening_gate#raw in
      ErrorHandling.raise_if_error ();
      result
    )

end